### DISCUSSION TITLE
Define a textual format for Schematron (#22)

### DISCUSSION THREAD

**Discussion by tgraham-antenna** [Schematron/schematron-enhancement-proposals#23] [Ideas]
Discussing #22...

From https://github.com/Schematron/schematron-enhancement-proposals/issues/22#issuecomment-927007028:

> I like simplified formats, but they can be a "if you only have a hammer,
> everything is a nail" solution: is the real issue formats, or
> familiarity/interaction modes for non-pointy people?  Here are two
> alternative suggestions:
>
> *1.  SPREADSHEET:  *
...
> *2. DSL/NL interface*

From https://github.com/Schematron/schematron-enhancement-proposals/issues/22#issuecomment-927595767:

> Maybe something along the lines of RelaxNG with its text format?

RelaxNG's compact syntax is the obvious comparison. I didn't want to ask about a 'Schematron compact syntax', however. If compactness was the goal, you could make something that's compact but unreadable (not that we would) and still claim success.

Maybe the answer to "Is it time for a textual format for Schematron?" really is "No, but it is time for a standard spreadsheet." However, I'd be more likely to use a textual format than a spreadsheet if both were available.

Rewriting XPath as COBOL seems like a step beyond an easier-to-read-and-write syntax for Schematron, particularly when you jump straight to describing a software system that responds as people type. An implementation of that might be just what people want, but I find it unlikely that the people who'd want that would want to write the NL syntax into attribute values in raw XML.

People are generally quick to understand XPath concepts in XSLT training sessions, and conversations with people who've taught Schematron courses indicates that people on the editorial side can grok Schematron quickly.

We sometimes go to great lengths to hide the tags in editing software interfaces, but we're still requiring people to see all the tags when creating Schematron files.  That doesn't make much sense to me. A way to round-trip between text and XML for Schematron might make Schematron easier to use for some people.

What this discussion really needs, however, is input from non-technologists about what would work best for them.

**Comment by PullFrom** [comment]
Well, before we switched to ISO Schematron we had a smaller private implementation of something very similar but editable in a regular XML editor. 
* All tests were written as element content, the messages also, of course
* Instead of the distinction between `<report>` and `<assert>` we used `<if-true>` and `<if-false>` child elements
* And it was possible to have additional test in the `<if-true>`/`<if-false>` branches, basically drilling down recursively

I proposed this as a talk in Prague, but it was rejected ("Why should one invent something like Schematron if we already have Schematron").
Technically the file was processed using XSLT to create an XSLT stylesheet for the actual tests.
The main benefit was the simpler editing without the need of a full-fledged XML IDE.

**Reply by tgraham-antenna** [reply]
> I proposed this as a talk in Prague, but it was rejected ("Why should one invent something like Schematron if we already have Schematron").

Heresy!

> The main benefit was the simpler editing without the need of a full-fledged XML IDE.

Yes. The point of #22 (sorry again for the duplication) is to see if there is a simpler way to read and write Schematron that makes Schematron easier to use (and broadens its appeal).

I have been doing markup for decades, but I would still rather use RelaxNG compact syntax than RelaxNG's XML format.

**Reply by dmj** [reply]
@michael-aka-mmh I started to study Schematron-like systems and would be eager to know more about your private implementation. I'm working on a small intermediate language that implements rule-based validation as a thin layer over XSLT (working title: The XSLT Schematron Framework). I've implemented ISO Schematron, XCSL, CLiX (partially) so far. Would be interesting to see how/if I could implement the private implementation as well and see what it could do that ISO Schematron can't.

**Reply by PullFrom** [reply]
Hi David, we have now removed all instances of this implementation and moved to ISO Schematron. I attach a small sample file which shows the recursive nature of rules. Using a very simple XSLT build stylesheet this was transformed into an XSLT which did the actual testing. HTH, - Michael
[otherprops-validation.zip](https://github.com/Schematron/schematron-enhancement-proposals/files/7596916/otherprops-validation.zip)

**Comment by rjelliffe** [comment]
I think Tony touched on this, but very often people are frustrated by a
particular syntax because it has too much fluff for their use-case.  The
mental effort figuring out that some aspect of the language (i.e.
Schematron here) is not relevant, of attempting shoehorning, can be more
taxing that figuring out what the parts that are good fits.

For example, what if you actually are not validating a document, but doing
feature extraction?   You can certainly use Schematron, with sch:report and
@role, and then check the SVRL for those roles.  But there is a lot of
unnecessary syntax, and allowing any kind of chaining may be horrible.
Feature extraction is certainly find the "presence or absence of XPaths in
an XML document" and therefore suited for Schematron, but what we are
interested in is the feature identifier to do some computer processing, not
an explanation for humans, necessarily.

In which case, perhaps the answer in that kind of case is not an easier
syntax for Schematron, but another little language? More direct. Surely one
lesson of Schematron is that little general but targetted languages that
rely on the power of XPath have some value?

Cheers
Rick

In the case of feature extraction, I have made a language for this purpose:
Feature Grammars has a more direct syntax and output format, once you
understand what it is trying to do: reduce the input document to a small
summary document that allows classification/dispatching (for example, to
detect that the document should be processed by one pipeline not another,
or even to then choose the appropriate Schematron phase for validation.)
See
    https://schematron.com/document/386.html
with better explanation at

https://bitbucket.org/CharmOnset/featuregrammars/raw/dfe85d94e63d79a2c432eff62f5cb5a9b2c34115/doc/FeatureGrammars-2016-10-10.pdf
)
